utility/importer/chrome_importer.cc

```diff
@@ -38,8 +38,78 @@
 #include "ui/base/l10n/l10n_util.h"
 #endif  // defined(OS_LINUX)
 
+#if defined(OS_WIN)
+#include "base/base64.h"
+#include "base/win/wincrypt_shim.h"
+#endif
+
 using base::Time;
 
+namespace {
+
+// Most of below code is copied from os_crypt_win.cc
+#if defined(OS_WIN)
+// Contains base64 random key encrypted with DPAPI.
+const char kOsCryptEncryptedKeyPrefName[] = "os_crypt.encrypted_key";
+
+// Key prefix for a key encrypted with DPAPI.
+const char kDPAPIKeyPrefix[] = "DPAPI";
+
+bool DecryptStringWithDPAPI(const std::string& ciphertext,
+                            std::string* plaintext) {
+  DATA_BLOB input;
+  input.pbData =
+      const_cast<BYTE*>(reinterpret_cast<const BYTE*>(ciphertext.data()));
+  input.cbData = static_cast<DWORD>(ciphertext.length());
+
+  DATA_BLOB output;
+  BOOL result = CryptUnprotectData(&input, nullptr, nullptr, nullptr, nullptr,
+                                   0, &output);
+  if (!result) {
+    PLOG(ERROR) << "Failed to decrypt";
+    return false;
+  }
+
+  plaintext->assign(reinterpret_cast<char*>(output.pbData), output.cbData);
+  LocalFree(output.pbData);
+  return true;
+}
+
+// Return false if encryption key setting is failed.
+// Fetch chrome's raw encryption key and use it to get chrome's password data.
+bool SetEncryptionKeyForPasswordImporting(
+    const base::FilePath& local_state_path) {
+  std::string local_state_content;
+  base::ReadFileToString(local_state_path, &local_state_content);
+  base::Optional<base::Value> local_state =
+      base::JSONReader::Read(local_state_content);
+  if (auto* base64_encrypted_key =
+          local_state->FindStringPath(kOsCryptEncryptedKeyPrefName)) {
+    std::string encrypted_key_with_header;
+
+    base::Base64Decode(*base64_encrypted_key, &encrypted_key_with_header);
+
+    if (!base::StartsWith(encrypted_key_with_header, kDPAPIKeyPrefix,
+                          base::CompareCase::SENSITIVE)) {
+      return false;
+    }
+    std::string encrypted_key =
+        encrypted_key_with_header.substr(sizeof(kDPAPIKeyPrefix) - 1);
+    std::string key;
+    // This DPAPI decryption can fail if the user's password has been reset
+    // by an Administrator.
+    if (DecryptStringWithDPAPI(encrypted_key, &key)) {
+      OSCrypt::SetRawEncryptionKey(key);
```

Comment:
Haven't looked into it but is there a danger here of overwriting an existing encryption key since this is a static?

Reply:
Agreed. This could overwrite if upstream explicitly sets encryption key to profile importer process.
Currently, no one set the key. but don't know upstream could set it in the future.
To make more safe, how about below steps during the password importing?
```
Import Password
   Cache current key
   Set chrome's encryption key
   Do importing
   Restore cached key
```

Reply:
Upstream code doesn't allow overwrite encryption key.
We can detect the overwrite if our code overwrite.
How about going with current PR?
If upstream code sets encryption key in importer service process, we can handle it in the future.
```
void OSCrypt::SetRawEncryptionKey(const std::string& raw_key) {
  DCHECK(!g_use_mock_key) << "Mock key in use.";
  DCHECK(!raw_key.empty()) << "Bad key.";
  DCHECK(GetEncryptionKeyFactory().empty()) << "Key already set.";
  GetEncryptionKeyFactory().assign(raw_key);
}
```